src/Crypto.gren

```diff
@@ -0,0 +1,2227 @@
+- `extractable` denotes that if this key is extractable or not. For more information,
+see the [`Extractable` type documentation](#Extractable).
+
+A missing part of these parameters is the public exponent. Generated keys will always
+have a `[ 1, 0, 1 ]` or `65537` public exponent. This is a recommended value and a value 
+that works across the `browser` and `node` platforms. If you have need for other public
+exponents, please file a ticket!
+-}
+type alias RsaKeyParams =
+    { modulusLength : Int
+    , hash : DigestAlgorithm
+    , extractable : Extractable
+    }
+
+
+{-| Errors that can happen when generating a key for use with the RSA algorithm. There is
+a single case where this function can fail at runtime:
+
+- When the passed `modulusLength` is not divisible by 8, as is required by the algorithm.
+This is captured by `ModulusLengthNotDivisibleByEight`.
+-}
+type RsaKeyGenerationError
+    = ModulusLengthNotDivisibleByEight
+
+
+{-| Generate a new key pair using the RSA-OAEP algorithm.
+
+Produces a `KeyPair` that can be used to encrypt data with [`encryptWithRsaOaep`](#encryptWithRsaOaep) 
+and decrypt data with [`decryptWithRsaOaep`](#decryptWithRsaOaep).
+-}
+generateRsaOaepKeyPair : RsaKeyParams -> SecureContext -> Task RsaKeyGenerationError (KeyPair RsaOaepKey RsaKeyParams)
+generateRsaOaepKeyPair params _context =
+    generateRsaKeyHelper 
+        "RSA-OAEP" 
+        [ "encrypt", "decrypt" ]
+        params
+
+
+{-| Generate a new key using the RSA-PSS algorithm.
+
+Produces a `KeyPair` that can be used to sign data with [`signWithRsaPss`](#signWithRsaPss) 
+and verify data with [`verifyWithRsaPss`](#verifyWithRsaPss).
+-}
+generateRsaPssKeyPair : RsaKeyParams -> SecureContext -> Task RsaKeyGenerationError (KeyPair RsaPssKey RsaKeyParams)
+generateRsaPssKeyPair params _context =
+    generateRsaKeyHelper 
+        "RSA-PSS"
+        [ "sign", "verify" ]
+        params
+
+
+{-| Generate a new key using the RSASSA-PKCS1-v1_5 algorithm.
+
+Produces a `KeyPair` that can be used to sign data with 
+[`signWithRsaSsaPkcs1V1_5`](#signWithRsaSsaPkcs1V1_5) and verify data with 
+[`verifyWithRsaSsaPkcs1V1_5`](#verifyWithRsaSsaPkcs1V1_5).
+-}
+generateRsaSsaPkcs1V1_5KeyPair : RsaKeyParams -> SecureContext -> Task RsaKeyGenerationError (KeyPair RsaSsaPkcs1V1_5Key RsaKeyParams)
+generateRsaSsaPkcs1V1_5KeyPair params _context =
+    generateRsaKeyHelper 
+        "RSASSA-PKCS1-v1_5"
+        [ "sign", "verify" ]
+        params
+
+
+{-|-}
+generateRsaKeyHelper : String -> Array String -> RsaKeyParams -> Task RsaKeyGenerationError a
+generateRsaKeyHelper name permissions { modulusLength, hash, extractable } =
+    let
+        clampedModulusLength = 
+            clamp 2048 4096 modulusLength
+    in
+    if Math.remainderBy 8 clampedModulusLength == 0 then
+        Gren.Kernel.Crypto.generateRsaKey 
+            name
+            clampedModulusLength
+            [ 1, 0, 1 ] -- Public exponent
```

Comment:
This should always be `[ 1, 0, 1 ]` ?

Reply:
@robinheghan In the [Web Crypto API docs](https://developer.mozilla.org/en-US/docs/Web/API/RsaHashedKeyGenParams), it's the recommended value. The reason it's hard-coded is I had a lot of trouble determining what is a allowable value for this and signaling the proper error. Oh, and if I recall, behavior across browser was pretty inconsistent. We can open this as a parameter that can be passed to RSA key generation functions, but if the value is incorrect, it'll still throw a generic error. It's not really possible, from what I tested, to tell them the public exponent specifically is invalid.

Reply:
Fair enough. Leave it the way it is 👍